File: leetcode/python/000090.subsets-ii.py
class Solution:
    # Time complexity: O(2^n)
    # Space complexity: O(n)
    def subsetsWithDup(self, nums: list[int]) -> list[list[int]]:
        res = []
        nums.sort()

        def backtrack(container, index):
            if index == len(nums):
                res.append(container.copy())
                return

            container.append(nums[index])
            backtrack(container, index + 1)

            container.pop()
            while index < len(nums) and index + 1 == nums[index]:
                index += 1
            backtrack(container, index + 1)

        backtrack([], 0)

        return res

File: leetcode/typescript/000090.subsets-ii.ts
// Time complexity: O(2^n)
// Space complexity: O(n)
function subsetsWithDup(nums: number[]): number[][] {
  const res: number[][] = [];
  const temp: number[] = [];
  nums.sort();

  function dfs(i: number) {
    if (i === nums.length) {
      res.push([...temp]);
      return;
    }

    temp.push(nums[i]);
    dfs(i + 1);

    temp.pop();
    while (i < nums.length && nums[i] === nums[i + 1]) {
      i++;
    }
    dfs(i + 1);
  }

  dfs(0);
  return res;
}
